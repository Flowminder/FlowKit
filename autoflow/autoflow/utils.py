# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
Utility functions.
"""

import collections
import json
import nbconvert
import nbformat
import networkx as nx
import pendulum
from contextlib import contextmanager
from get_secret_or_env_var import getenv
from hashlib import md5
from pathlib import Path
from sh import asciidoctor_pdf
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from tempfile import TemporaryDirectory
from typing import Any, Dict, Iterator, Optional, OrderedDict, Set, Tuple, Union


def get_output_filename(input_filename: str, tag: str = "") -> str:
    """
    Given an input filename, construct an output filename with the same extension,
    with a timestamp (and optionally a tag) added to the stem.

    Parameters
    ----------
    input_filename : str
        Input filename
    tag : str, optional
        A tag to append to the file stem
    
    Returns
    -------
    str
        Output filename
    """
    input_filename = Path(input_filename)
    now_string = pendulum.now("utc").format("YYYYMMDDTHHmmss[Z]")
    tag = f"__{tag}" if tag != "" else tag
    return f"{input_filename.stem}{tag}__{now_string}{input_filename.suffix}"


def get_params_hash(parameters: Dict[str, Any]) -> str:
    """
    Generate a md5 hash string from a dictionary of parameters.
    The dictionary is dumped to a json string before hashing.

    Parameters
    ----------
    parameters : dict
        Dictionary of parameters to use for creating the hash.
    
    Returns
    -------
    str
        md5 hash of the parameters dict
    """
    return md5(
        json.dumps(dict(parameters), sort_keys=True, default=str).encode()
    ).hexdigest()


def get_session(db_uri: str) -> "sqlalchemy.orm.session.Session":
    """
    Create a sqlalchemy session.

    Parameters
    ----------
    db_uri : str
        Database URI
    
    Returns
    -------
    Session
        A sqlalchemy session
    """
    # TODO: This seems like the wrong place to be reading a secret / env var,
    # but we can't put a docker secret in the prefect config.
    full_db_uri = db_uri.format(getenv("AUTOFLOW_DB_PASSWORD", ""))
    engine = create_engine(full_db_uri)
    return sessionmaker(bind=engine)()


@contextmanager
def session_scope(db_uri: str) -> Iterator["sqlalchemy.orm.session.Session"]:
    """
    Provide a scoped sqlalchemy session for interacting with the database.

    Parameters
    ----------
    db_uri : str
        Database URI
    
    Yields
    ------
    Session
        A sqlalchemy session
    """
    session = get_session(db_uri)
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()


def make_json_serialisable(obj: Any) -> Union[dict, list, str, int, float, bool, None]:
    """
    Helper function to convert an object's type so that it can be serialised using
    the default json serialiser. Non-serialisable types are converted to strings.

    Parameters
    ----------
    obj : any
        Object to be converted.
    
    Returns
    -------
    dict, list, str, int, float, bool or None
        The same object after dumping to json (converting to str if necessary) and loading
    """
    return json.loads(json.dumps(obj, default=str))


def get_additional_parameter_names_for_notebooks(
    notebooks: Dict[str, Dict[str, Any]],
    reserved_parameter_names: Optional[Set[str]] = None,
) -> Set[str]:
    """
    Extract parameter names from a list of notebook task specifications.

    Parameters
    ----------
    notebooks : dict
        Dictionary of dictionaries describing notebook tasks.
    reserved_parameter_names : set of str, optional
        Names of parameters used within workflow.
    
    Returns
    -------
    set of str
        Names of parameters used by notebooks which must be passed to a workflow.
    """
    # Parameters requested in notebooks
    notebook_parameter_names = set.union(
        *[set(notebook["parameters"].values()) for notebook in notebooks.values()]
    )
    # Parameters available to notebooks, which are not passed externally as flow parameters
    internal_parameter_names = (reserved_parameter_names or set()).union(
        notebooks.keys()
    )
    # Additional parameters required for notebooks
    additional_parameter_names_for_notebooks = notebook_parameter_names.difference(
        internal_parameter_names
    )
    return additional_parameter_names_for_notebooks


def sort_notebooks(
    notebooks: Dict[str, Dict[str, Any]]
) -> OrderedDict[str, Dict[str, Any]]:
    """
    Perform a topological sort on a dictionary of notebook task specifications.

    Parameters
    ----------
    notebooks : dict
        Dictionary of dictionaries describing notebook tasks.
    
    Returns
    -------
    OrderedDict
        Ordered dict of notebook task dicts, ordered so that no notebook depends on another
        notebook that comes after it.
    
    Raises
    ------
    ValueError
        If the notebook specifications contain circular dependencies.
    """
    notebooks_graph = nx.DiGraph(
        {
            key: [
                value
                for value in notebooks[key]["parameters"].values()
                if value in notebooks
            ]
            for key in notebooks
        }
    ).reverse()
    try:
        sorted_notebook_keys = list(nx.topological_sort(notebooks_graph))
    except nx.NetworkXUnfeasible:
        raise ValueError("Notebook specifications contain circular dependencies.")

    return collections.OrderedDict(
        (key, notebooks[key]) for key in sorted_notebook_keys
    )


def notebook_to_asciidoc(
    notebook_path: str, asciidoc_template_path: Optional[str] = None
) -> Tuple[str, "nbconvert.exporters.exporter.ResourcesDict"]:
    """
    Convert a Jupyter notebook to ASCIIDoc, using nbconvert.

    Parameters
    ----------
    notebook_path : str
        Path to notebook
    asciidoc_template_path : str, optional
        Path to a template file to use for the conversion.
        If not provided, the default nbconvert asciidoc template will be used.

    Returns
    -------
    str
        The resulting converted notebook.
    ResourcesDict
        Dictionary of resources used prior to and during the conversion process.
    """
    with open(notebook_path) as nb_file:
        nb_read = nbformat.read(nb_file, as_version=4)

    if asciidoc_template_path is None:
        exporter = nbconvert.ASCIIDocExporter()
    else:
        exporter = nbconvert.ASCIIDocExporter(template_file=asciidoc_template_path)
    body, resources = exporter.from_notebook_node(nb_read)

    return body, resources


def asciidoc_to_pdf(
    body: str, resources: "nbconvert.exporters.exporter.ResourcesDict", output_path: str
) -> None:
    """
    Convert an ASCIIDoc document to PDF, using asciidoctor-pdf.

    Parameters
    ----------
    body : str
        Content of the ASCIIDoc document
    resources : ResourcesDict
        Dictionary of resources, as returned from nbconvert exporter
    """
    with TemporaryDirectory() as tmpdir:
        with open(f"{tmpdir}/tmp.asciidoc", "w") as f_tmp:
            f_tmp.write(body)
        for fname, content in resources["outputs"].items():
            with open(f"{tmpdir}/{fname}", "wb") as fout:
                fout.write(content)
        asciidoctor_pdf(f"{tmpdir}/tmp.asciidoc", o=output_path)
