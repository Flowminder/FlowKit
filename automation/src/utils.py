from pathlib import Path

def get_output_filename(input_path, output_dir, label):
    input_path = Path(input_path)
    return f"{output_dir}/{input_path.stem}_{label}{input_path.suffix}"