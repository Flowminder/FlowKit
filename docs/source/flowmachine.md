# FlowMachine

FlowMachine is a Python toolkit for the analysis of CDR data. It is essentially a python wrapper for postgres SQL queries, including geographical queries with postgis. All of the classes in flowmachine define an SQL query under the hood.

## Documentation

Documentation for FlowMachine can be found [here](../developer/api/flowmachine).
