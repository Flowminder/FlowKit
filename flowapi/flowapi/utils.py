# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.


from quart import request
from quart.exceptions import HTTPException


async def get_query_parameters_from_flowmachine(*, query_id) -> dict:
    """
    Get the parameters of a query from flowmachine.

    Parameters
    ----------
    query_id : str
        ID of the query to get params for

    Returns
    -------
    dict
        Dictionary containing the query's original parameters

    Raises
    ------
    HTTPException
        404 if the query id is not known.

    """
    request.socket.send_json(
        {
            "request_id": request.request_id,
            "action": "get_query_params",
            "params": {"query_id": query_id},
        }
    )
    reply = await request.socket.recv_json()
    if reply["status"] == "error":
        raise HTTPException(
            description=f"Unknown query ID '{query_id}'",
            name="Query ID not found",
            status_code=404,
        )
    return reply["payload"]["query_params"]
