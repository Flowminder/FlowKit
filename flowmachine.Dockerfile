# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

FROM python:3.8-slim

ARG SOURCE_VERSION=0+unknown
ENV SOURCE_VERSION=${SOURCE_VERSION}
ENV SOURCE_TREE=FlowKit-${SOURCE_VERSION}
WORKDIR /${SOURCE_TREE}/flowmachine
COPY ./flowmachine/Pipfile* ./
RUN apt-get update && \
        apt-get install -y --no-install-recommends git && \
        pip install --no-cache-dir pipenv && pipenv install --clear --deploy && \
        apt-get -y remove git && \
        apt purge -y --auto-remove && \
        rm -rf /var/lib/apt/lists/*
COPY . /${SOURCE_TREE}/
RUN apt-get update && \
        apt-get install -y --no-install-recommends git && \
        pipenv run python setup.py install && \
        apt-get -y remove git && \
        apt purge -y --auto-remove && \
        rm -rf /var/lib/apt/lists/*
CMD ["pipenv", "run", "flowmachine"]
# FlowDB has a default role named flowmachine for use with the flowmachine server
# when starting the container with a different user, that user must be in the flowmachine
# role
ENV FLOWMACHINE_FLOWDB_USER=flowmachine
