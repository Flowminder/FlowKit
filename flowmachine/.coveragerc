[run]
branch = True

[report]
fail_under = 90

omit =
    */python?.?/*
    */venv/*
    */site-packages/nose/*
    */tests/*
    */_version.py
    */versioneer.py
    *__init__*
    flowmachine/core/server/server.py

exclude_lines =
    if __name__ == '__main__':

show_missing = True
