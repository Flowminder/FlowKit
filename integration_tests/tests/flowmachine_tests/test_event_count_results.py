# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from flowmachine.utils import pretty_sql

from approvaltests.approvals import verify
from flowmachine.core import CustomQuery
from flowmachine.features import EventCount


def test_event_count_1_sql(diff_reporter):
    """
    Event count query with non-default parameters returns the expected data.
    """
    ec = EventCount(start="2016-01-02", stop="2016-01-04", tables=None)
    sql = pretty_sql(ec.get_query())
    verify(sql, diff_reporter)


def test_event_count_1_df(get_dataframe, diff_reporter):
    """
    Event count query with non-default parameters returns the expected data.
    """
    ec = EventCount(start="2016-01-02", stop="2016-01-04", tables=None)
    df = get_dataframe(ec)
    verify(df.to_csv(), diff_reporter)
