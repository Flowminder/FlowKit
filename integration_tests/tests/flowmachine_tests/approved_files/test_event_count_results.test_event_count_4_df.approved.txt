,subscriber,value
0,09NrjaNNvDanD8pk,2
1,0DB8zw67E9mZAPK2,1
2,0Gl95NRLjW2aw8pW,4
3,0MQ4RYeKn7lryxGa,6
4,0NVaL8k1nwRwOmlg,1
5,0W71ObElrz5VkdZw,3
6,0Ze1l70j0LNgyY4w,2
7,0ayZGYEQrqYlKw6g,2
8,0gmvwzMAYbz5We1E,1
9,0zqZPERbraE8NM4W,3
10,148ZaRZe54wPGQ9r,3
11,1NqnrAB9bRd597x2,1
12,1QBlwRo4Kd5v3Ogz,2
13,1d29oEA95KEzAKlW,1
14,1jwYL3Nl1Y46lNeQ,1
15,1p4MYbA1Y4bZzBQa,1
16,1vGR8kp342yxEpwY,2
17,24ndeQw41nGqlPZ1,5
18,2Dq97XmPqvL6noGk,4
19,2GJxeNazvlgZbqj6,1
20,2MdDGpm37Dr41nAY,1
21,2NPz1Vz1AaKDLoGQ,2
22,2PjYkGv6ARGqQ67p,4
23,2ZRmOxEWBVq546zY,6
24,2ZdMowMXoyMByY07,5
25,2dpYQW4pzMWBxORA,4
26,2vmOlAENnxpPM1xX,2
27,2yqlaY6LD9Q8xZj1,1
28,346j7Nq67nvXZR0m,2
29,37J9rKydzJ0mvo0z,1
30,3DnLz7g07mkQepaO,4
31,3EgqzplqPYDyGRVK,2
32,3GZ48OAlG6mPMBKg,3
33,3XKdxqvyNxO2vLD1,2
34,3enxoqbomNMB97lW,2
35,3o9YZlRpm74nwDXx,4
36,3x5BKX29EzQVDbJy,4
37,4587JeOdRBEW96K1,1
38,4AqjP2LWpw3r73vB,4
39,4PWQE7lNA25NYV6p,1
40,4QVGRaVGwydYE86n,3
41,4QyYADMYXqPDP79G,2
42,4ZVwldpn4QXDN0OY,4
43,4ZpLvnNZArnNae9R,4
44,4dqenN2oQZExwEK2,1
45,4g7a3XELre5dYzr6,1
46,4oLKbnxm3vXqjMVx,5
47,58nJwlYY8El6VX0E,4
48,5KaRV2B36vpNoybG,7
49,5Kgwy8Gp6DlN3Eq9,2
50,5Q0lndPxZYdZGMO8,2
51,5jLW0EWeoyg6NQo3,2
52,5oDLGWg9lBeQRP9d,4
53,5q2VK37925edBXb8,2
54,5vwqabdVV1lZyLBE,1
55,5xZ8YkbRqrkv1r4J,1
56,5yPVlxl9aDbK2L0r,4
57,62QyANzLydvMzVgm,1
58,6ALdEroGZZrjqbgl,4
59,6NKDrZDNWwayGO9j,1
60,6NWOjmXPkxmbaw73,2
61,6kj8RD7YQ4kwWlQy,3
62,6nYkgBqZNo5KG7N4,1
63,6qMK0QbqlAnWl2dN,4
64,6v9YyNx3r7zb3kO8,3
65,6wGDKd79v0bnMzor,2
66,6zm4b7veY2ljEQ95,3
67,7JyRXDKdd4gnwq0d,2
68,7XebRKr35JMJnq8A,1
69,7b4BZo0kwQYRmyJ0,2
70,7dePlr8JjqRQ9vGz,4
71,7lNP0mDOAK3xKWv4,1
72,7nWrYGEd029zJ02M,1
73,7nqEVPQ06ZkQ9w3a,1
74,7qKmzkeMbmk5nOa0,1
75,7ra3xZakjEqB1Al5,2
76,7wmpbZzKRnaR4GvO,1
77,81B6q0K8k325XWmn,2
78,83Kywz7PZqjMPJpR,3
79,87mobXjNNDVX51ML,6
80,8DLbw958mPv3xjVa,3
81,8GkR7zbjLezebY5E,1
82,8JeRLGmxlq0k0qZz,2
83,8Nerw0xj58Ky5RVz,3
84,8b6qy2aDVmjlaogE,4
85,8bGZXqJXP1Jgr1EM,3
86,8br1gO32xWXxjY0R,3
87,8dXPM6JXj05qwjW0,4
88,8dpPLR15XwR7jQyN,3
89,8e4zNGWqVArxE2Zo,1
90,8lo9EgjnyjgKO7vL,3
91,8rxEQOL3ePZdAe1z,2
92,951zd63qryZAYBy7,6
93,9eDMvX85yLR3dGlE,4
94,9vXy462Ej8V1kpWl,2
95,9wkjq8LdrAr84Y6g,4
96,A42Xl6GzjBqMODyz,3
97,A8Y0ylLjaDDQwORD,3
98,ABV1RygG9oyXoJ8j,2
99,AMjGX58pWv5Ql2o6,1
100,APj9roe8jKOwEDZl,1
101,AZj6MqBAryVyNRDo,2
102,Aa3kjzJmqZWveXqb,4
103,AgB6KR3Levd9Z1vJ,3
104,AgvE8pa3Bvqezmo6,2
105,Ak4LV2l6yNDrg9qw,2
106,Ap3E6Ljl0kWRavy4,3
107,Aqp2DjeEmwjEVZKo,2
108,AwQr1M3w8mMvV3p4,1
109,B4JG213EnO0bwXAN,3
110,B7blVrVn0br1zP3x,4
111,BGXbdrXG9J2gVZoA,2
112,BKMy1nYEZpnoEA7G,1
113,BNxWrJ9mqMJ83Mzk,3
114,BVYqp0ryO1oj1gRo,2
115,Bar6kR1QXKRlYn1J,3
116,BebaVL8XZzLn18y2,4
117,Bn5kZrQ2WgEy14zN,9
118,Bp8QkGlv3E4mxvOP,6
119,Bwo5pgJzV1ogxyeK,2
120,By4j6PKdB4NGMpxr,4
121,D6b8NwVBmmw5JzA1,3
122,D8AZvkA5NXmkyEeO,4
123,D9qXvyM1NbyxpdEm,3
124,DAlqeZENbeOn2vBw,3
125,DELmRj9Vvl346G50,6
126,DWV9mxjjvLxwnXyb,4
127,DpmEKQaW8X3PdZLo,5
128,DzpZJ2EaVQo2X5vM,3
129,E1n7JoqxPBjvR5Ve,3
130,E9nlO1dMyVAWr60X,1
131,EYojLpzGVbzyJQvn,2
132,Ea2DjkNvAvkp7OMe,3
133,EapG7nvNNpn1PAdv,2
134,EdWZO2dVLNRDMPlV,4
135,Egly7ZQvwdopGmYP,1
136,EkpjZe5z37W70QKA,3
137,EqrX8mBXDZPMRKLz,1
138,EreGoBpxJOBNl392,2
139,G0VxgAwoOPAM685Q,6
140,G2DQzae1qOa48jK9,5
141,GMKmXnYNgVRlyb6r,5
142,GNLM7eW5J5wmlwRa,1
143,GNmpb80ZawxQ7WOV,3
144,GXd06jGNEamgBwr3,4
145,Gd2bxpQG0xR91VXD,2
146,GnyZMedmKQ4X78Wa,3
147,J0Yyqw2rkVEwpMG2,3
148,J4K6gkq7wyqopOnQ,6
149,J7gz8x0KJKn5y3MG,5
150,JAqlnPX1QmZ8pQ5V,4
151,JN6WPrJPeKV1lx3q,5
152,JNK7mk5G1Dy6M2Ya,3
153,JO3QBk9PmjoMmge6,1
154,JRZMbq2eOM189Qw4,4
155,JZoaw2jzvK2QMKYX,3
156,JrYLpWRmryNm0xDP,5
157,KAMQan1jjab4rlgX,1
158,KAygLZgx8oAeJz8Y,3
159,KD1ro51Yo6o5vg6n,5
160,KM4YzL5aPRL59dRy,1
161,Kabe7EppYbMEj6O3,4
162,KmdyQlgVyQjDWvJN,3
163,Kxb2JoGX50lzRDEq,1
164,Kxj9rpMEA9pMmV8J,1
165,KyNxbYLE2Qzj3mPD,3
166,KyonmeOzDkEN40qP,4
167,L4V537alj321eWz6,4
168,LBlWd64rqnMGv7kY,3
169,LR6Dn0DNJeGx39g8,3
170,Le5bxBw5VpqZpGQ2,1
171,LjDxeZEREElG7m0r,2
172,Ln4ly8DYR6wzWDMo,3
173,LnKG6xoRJ0X1wm7Y,1
174,LqgwYrR7ZZp5xGK1,2
175,M0pdYQAADkQXP9Vy,2
176,M60LdelQ6ROXRJ3P,1
177,M87PjXJVWroXROWw,2
178,M9YvRrkbpeZ052yj,2
179,MLo57YoM963gRQe1,1
180,MRwEk8RmvK2Y0P9N,1
181,MX0JR5y11e5narO6,3
182,MXLPGvyz63An4bE1,3
183,MZze6BkgGoBv5dEG,4
184,Ma0Xv1XJxE2JV7r8,1
185,Mmz20GVDbqjEel51,3
186,MobnrVMDK24wPRzB,6
187,N5vEaB5gO5yDxRrl,3
188,N640d1AGW8pyxWKn,1
189,N6R3Akx0zWk05oWp,4
190,NA7D19MmEm9gkK5R,4
191,NBpb54xVMEw1Yqza,1
192,NG1km5NzBg5JD8nj,2
193,NOXLME8YWo09J6oY,4
194,NQV3J52PeYgbLm2w,2
195,Nbng985Vjoa0Ze1X,4
196,NeRzMVGy2MEvgjDZ,3
197,NejWvxnkxjlXpJgQ,2
198,Nnlqka1oevEMvVrm,2
199,NvzxDwNZo0wleBg5,2
200,NzlPGkByB5jqBJ1v,3
201,O2oGlMjNX7MpwV43,1
202,O8a7ADMbbmv0QkbV,5
203,OBwVW2WZNdMrEv7Q,1
204,OdM7np8LYEp1mkvP,1
205,OemQ7q2DLZMWnwzB,5
206,Oko9NEkexzJQKReX,2
207,Oxv502qrde8KmgVr,5
208,OzlrV3GEOyn5DW1Y,7
209,P5Bwzx72E80Dno21,1
210,P7kMZxeEOzxr2Wg1,2
211,P8RpJMqGpexyOl2o,1
212,PJ536E18n1D4Bd1r,2
213,PR31Yr7agDbJzaZX,4
214,PWlZ8Q5zjYkGJzXg,6
215,PZVjpGbPvyjgQErM,2
216,PagZ0KRaVNlY58kQ,4
217,PbN9vQlbEORLmaxd,2
218,PeJKXdZRoadyEMaj,1
219,PnxGEed4JWezYdBg,5
220,Pq0xpgkyDrY6kmDy,4
221,PzQk5poRMA7ERaBe,2
222,Q0ELoRGG3X9Rg1GA,2
223,Q1jMk7qjqXBnwoDR,4
224,Q3xOLdlXJ4dMB7R1,2
225,Q4mwVxpBOo7X2lb9,1
226,Q7lRVD4YY1xawymG,3
227,QDqWpen1qakOLzd5,5
228,QEoRM9vlkV18N4ZY,2
229,QK78nqVO0o21VGB0,2
230,QPdr2B94VaEzZgoW,5
231,QWAjL9Nbyx95moq7,4
232,QWZzERBdXVM0GKXd,2
233,QWjwV5Xr3zEeKoDP,2
234,QeREg30jMjAd0Vz4,3
235,QoD6rWNN03WPbEBg,2
236,QrAlXqDbXDkNJe3E,5
237,RB4gWNwl4D8exP7l,1
238,RDqQa5nOVoX07Lk4,2
239,RGkzrL0JYKb2gPqW,4
240,RZkbX2rypqWja9Bv,3
241,RkdLwPWLboQ2WvX4,1
242,RnAv0YjPnW68x9wr,1
243,Rv8rPlM4wp54q7QZ,6
244,Rzx9WE1QRqdEX2Gp,1
245,V1QBpMj0vEwr2PGW,4
246,V4DJ7AaxLLOyWEZ3,2
247,V7MBRewnwQGE91gY,1
248,VGYEqRGVxVQl7JZO,1
249,Va0P7xJQVOmMpQn6,5
250,VgzRWn82MdlbXkyO,1
251,VkzMxYjv7mYn53oK,5
252,Vqjp2yxLk8yebXmP,3
253,Vqvyp1AGr7m58D4x,3
254,Vqy0OnN9P6lYgAze,4
255,W2VRAqvL4OM1xZEO,2
256,W6PRYNgyW89eZ8A1,3
257,W6ljak9BpW0Z7ybg,1
258,WKP7OnqzA2rxDA6J,5
259,WaekA0LbqwO6rpRG,5
260,WdBX5pVBE2pynxPQ,3
261,WkYAGXV6A1XOxg8o,6
262,Wm2PbGlxP7GAzvVn,4
263,X12gvGBdjp7E3Pbj,2
264,X2GyrjgBPNJpveaZ,3
265,X47N9O28wGd6Mneo,5
266,X809mynDXN621PoN,1
267,XQOrADNd8ePa71Ve,1
268,XRo0EdKqwD3zZBLr,2
269,XWRldepdNlegN3Zo,6
270,XjyWlpdZNZzY3O0N,3
271,Y8lvgqN0n75GQwk7,2
272,YGE2AwGPPEa86kq9,2
273,YK6z2lXzg7w57Vap,4
274,YMBqRkzbbxGkX3zA,2
275,YNv2EgDJxxAoy0Gr,4
276,Yd7BeVPEJA0xlJwr,3
277,YeqxBmVgL1EnONv8,2
278,YxDnjbQ4E9pVK4A6,1
279,Z4pBaoQLj4Gz76KA,3
280,Z51jagNxqkWR2L0o,2
281,Z89mWDgZrr3qpnlB,5
282,ZEjaJ0lkr19vYWRB,1
283,ZM3zYAPqx95Rw15J,1
284,ZQG8glazmxYa1K62,1
285,ZQbqE6Kw4d6WRnm2,3
286,ZYPxqVGLzlQy6l7n,1
287,ZeloJKzkEwVrLNYO,4
288,ZmPRjkQ74Xeql71V,1
289,ZnYwpaZxEjW6Ke7k,2
290,Zv4W9eak2QN1M5A7,5
291,a0ZQeXdY48Wg5bkw,2
292,a3PrBbGBAXpL8mp0,3
293,a76Ajyb9dmEYNd8L,4
294,a7wRAnjb2znjOlxK,1
295,aJmlqpLqaa8kn94P,1
296,aK9GDX8nozb2RB67,2
297,alK9yYKwJ3YgXqJW,2
298,alye4Z3yz5ENvnXY,4
299,amW6DwnDmbZ50xkO,2
300,aroL1Ar5K1jK9G26,2
301,b9Y3V7deAlgXolRd,4
302,bKZLwjrMQG7z468y,3
303,bMrnX9xrVNp0lvx2,4
304,bejv8KjzaA7K0319,3
305,bjY7mMXxE3zMoelO,1
306,dM4aLP8N97eYABwR,1
307,dWE82PWRRkX5wr0b,5
308,dXogRAnyg1Q9lE3J,4
309,dgz4WNZOAm5bmJ9M,2
310,dlyoVBZ5LR9J7QYW,3
311,dqEQWNwAYEr4Mk3e,3
312,dr9xNYK006wykgXj,5
313,dw0xVe5rw3gy14EL,4
314,e6BxY8mAP38GyAQz,1
315,eJ6vRwmbMEv5E7gq,4
316,eg5G9xM8aZAmnKp1,2
317,g9D5KEK9BQzOa8z0,2
318,gKkzQLa0yZ6Mvonm,3
319,gPDW4zr1YmVGqEAO,5
320,gPZ7jbqlnAXR3JG5,1
321,gQ0eDL57DgxKwqn8,2
322,gV57Wm5zvWdklYjz,3
323,gWPl7QBD8VnEyX8K,2
324,gWXN3Z5veAr162kw,3
325,gWwV6lpVON7ryzDe,4
326,gkBLe0mN5j3qmRpX,2
327,gwAynWXp4eWvxGP7,5
328,gxz7rWG7rVWj40vX,3
329,gz3o1Zr6yJZnOYr9,2
330,j6QYNbMJgAwlVORP,3
331,jERG2g84xGgVyz6K,2
332,jJOA5l9NyPEmVrPK,6
333,jNMBy0rVVd0DnGK6,5
334,jNQ9L7MjWLGebYM6,1
335,jPk9xAzO70rR1bBe,4
336,jW09JkYPjv6zmKP8,5
337,jWXpmeg35Ye3ByoJ,3
338,jWlyLwbGdvKV35Mm,3
339,jpXleRQqrLR0aPwN,1
340,jwKJorl0yBrZX5N8,1
341,jwgy7YympQYMBQVx,5
342,k8JRVodMrpKQp35q,2
343,k9J7lEbPxBnVyj3D,2
344,kJRMlxA71OL5om49,5
345,kYp5Xor8eGQ0DBZL,4
346,ka54Xb0N4x9LqPnE,3
347,kjGXLy9lWnZ4V6J7,4
348,kmR2eMYzRbAW8r4Q,2
349,kq7DLQar1nw2glYN,2
350,kw0rRZVr2vZB2zWl,3
351,l2RN3b7WRl2b1V8P,1
352,l6NQwr8qzwZLGy92,1
353,l7xALKEXZqjPaWQY,1
354,l81xEJB2XdJWNedP,3
355,l8KM7myrkvPLXpZx,4
356,lDv0jon4AyqL4byn,3
357,lGB9advjJYjdbWrN,3
358,lLVzNDy2BX1aO8wY,3
359,lPp65Z8gbEywqOnG,3
360,lRjYnM34X5AXBGrz,5
361,lVPvMWVBKOW74ZEO,4
362,lkPBK9BmJMpvMWnb,5
363,m4L326vrwE6elJxQ,4
364,m51bzWYMxKWLJy2V,2
365,mJ9eZYnvvr5YGW2j,1
366,mKyqZPM1AXPAQL08,1
367,mV1dOAOE0Vq4DLJk,2
368,mkqQ4NPBPQLapbeg,6
369,mpqkOe1L1wAaMRgE,7
370,mqP6d9k5nR21oaeb,2
371,nAELXw76mverkmqZ,4
372,nB4qp1mVza9bWY0E,4
373,nJLe9jzo96KVAXN7,4
374,nL9KYGXpz2G5mvDa,1
375,nLvm2gVnEdg7lzqX,6
376,nMvpK39bowVXYN9G,3
377,nRLeDl77Del5xpbY,1
378,nX89Y35pYvqAVvpD,2
379,ngRw8aKKdpaEmY3Q,5
380,nvKNoAmxMvBW4kJr,4
381,o1Rwj2XPYo2E05KP,1
382,o4ya5j92g5GRxdw2,1
383,o5V0lPmGoorXnv4q,2
384,oL13aVowaNpVk2np,3
385,oQOm1YkDxljp3AVL,3
386,oR78lLNzN9LpgqNb,2
387,oWkryLvyr0Jba1E5,3
388,oX6kEmV5yyb0xq3M,5
389,oX7ZVm05RPr9v0Aq,2
390,oanJmEDrLxpbVPzM,1
391,okNGOMXqlJjnRarA,1
392,onayVzV3pWrpW3M8,4
393,oq936AVWZRzYWPrZ,5
394,oqNR8gkbv6e4K97z,2
395,p6alg5rYl14BG2nL,2
396,p7N2YgEKM18vwmLe,4
397,p8oPjnDRq0M2wJW0,4
398,pB8g1y7jQEZO734d,1
399,pD0me6Do0dNE1nYZ,1
400,pDNARk60aQwO9Qar,3
401,pdrXzZj2DkZLJvbk,3
402,peMgdn594XnNO4bK,5
403,pnwa9PE5MbPB46O0,5
404,poXqlQOnjlJ5EZMj,3
405,pxgnQqP14BVkLM6o,2
406,q4rpo9aoJR3BbvWl,4
407,qJjd9VmvDrgQMA1O,2
408,qPAbaED3vDYkjZ0n,2
409,qPorDyvDD8RmgnL0,5
410,qRlQo7ly3Zg9GVN2,2
411,qj8gk6qbN8R3DBdO,2
412,r1av4JnM18dYGxAe,3
413,rG5oYMVaBj42Oa6L,2
414,rJeX25p8mn5Qwqmd,2
415,rmb4PG9gx1ZqBE01,3
416,v3kQyzZNRNAz5lbZ,1
417,vBR2gwl2B0nDJob5,2
418,vJK0DwMjZEm7gBQV,1
419,vJrRPQRm1lEma2KO,1
420,vKVLDx8koQWZ2ez0,2
421,vOGRpPpAArjBQZ9y,1
422,vWwlyp5a4GpdxmBZ,1
423,w1N59Jo07vbgrKae,1
424,w2YgmOPlr2O5x3VJ,1
425,wEZaqV15wMVGe1v6,4
426,wMbWOpJOk0GmZvLD,4
427,wX9ODedAaV5g8v26,4
428,wZReL3b1lEQkrm2G,6
429,wgoa5OApGQey32ZM,6
430,woOyJvzKBALYj8za,1
431,wrmEaJ9prNvyV4zL,1
432,wvMm1EAY0QeEAbN9,1
433,wynMvOdqX6Eo9PmR,5
434,wyrgMxBzzMxQVj83,3
435,x7wGEeYdBGD9aQ5X,2
436,x8o7D2Ax8Lg5Y0MB,2
437,xDQAPjbY0Z5o3Lqg,6
438,xE8QMJxmLGMW6lZ4,3
439,xJwr9glV6emG7b8o,2
440,xN4XQEa6JZJvkeAj,1
441,xkZb5E55LYE10wa4,1
442,xoJnvpa7LJNgyqKe,1
443,y2MZGq0N6370KXoz,3
444,yObw75JkAZ0vKRlz,3
445,yPzO2eJRrzeqnkEV,4
446,yYbzQ46yEx93D7je,2
447,ya9DBmDpJNm25lP1,2
448,ye8jQ0ovnGd9GlJa,1
449,yqQ6m96rp09dn510,3
450,yqw50eNyEwOxNDGL,1
451,z7mQgkVYoz2vr4EG,5
452,zGWn8opVmOQAD6xY,3
453,zj9KZblppy6byEDg,1
